samples/apps/promptflow-autogen/flow.dag.yaml

```diff
@@ -0,0 +1,31 @@
+$schema: https://azuremlschemas.azureedge.net/promptflow/latest/Flow.schema.json
+environment:
+  python_requirements_txt: requirements.txt
+inputs:
+  chat_history:
+    type: list
+    is_chat_history: true
+    default: []
+  question:
+    type: string
+    is_chat_input: true
+    default: Write a concise but engaging blogpost about programming language python
+  autogen_workflow_id:
+    type: int
+    default: 2
+outputs:
+  answer:
+    type: string
+    reference: ${autogen_task.output}
+    is_chat_output: true
+nodes:
+- name: autogen_task
+  type: python
+  source:
+    type: code
+    path: autogen_task.py
+  inputs:
```

Comment:
What if the user misses defining a few values in YAML, do we have proper error handling messages thrown ?

In all of these YAML files and json schema, how will the user know which variables can be skipped, and which of them need to be defined and are required ?

Reply:
Once this conversation is resolved, the PR will be merged.

Reply:
@krishnashed lets connect on Discord to discuss further

Summary: Promptflow support creating the steps using the DAG UI and is auto-converted to YAML by the framework itself - https://microsoft.github.io/promptflow/how-to-guides/quick-start.html#understand-the-flow-yaml
all the error handling for incorrect and invalid flow.dag.yaml is handled and managed by the promtpflow framework itself. User gets a visual display on where in the flow they need to correct the issue. Example screenshot below

simulating an error by referring to non-existent file:
![image](https://github.com/microsoft/autogen/assets/2416507/84f90b97-c955-4024-a364-56c1388f6d3a)

![image](https://github.com/microsoft/autogen/assets/2416507/94a56da3-bed7-4b85-a15d-13d2486f5508)

Reply:
Looks good , I wanted to double check the role of Redis - is it necessary? Or simply a cache example.

Secondly the output, is formatted to work as a promptflow result "output:<JSON>", you'll find it much easier to chain promptflows together if we follow that standard.

Reply:
Thanks @ElliotWood 
- Redis is just a cache example and i am planning to have a follow up on how to take those interim messages and return to the user. Autogen output generally take longer and i want to ensure we have a way to display the interim messages if required.

- In the current use-cases there is no need to have history and hence i did not want to over complicate. I am returning `ChatResult` object and getting `summary` field. You can get `chat_history` and other fields too.. need to change from `res.summary` to `res`. 
For the next PR i want to include Group chat with Human interaction and Include history and show end to end. Let me know your thoughts

![image](https://github.com/microsoft/autogen/assets/2416507/ca9e014a-efc0-43c9-911f-a7f78217d279)

![image](https://github.com/microsoft/autogen/assets/2416507/1f57ea3d-f562-44ba-92ce-d2927e56f94f)


![image](https://github.com/microsoft/autogen/assets/2416507/02b4a13c-0aae-49fa-8828-d6cbf06ded82)

Reply:
@krishnashed i have updated the readme as discussed and resolving this conversation as agreed. Thanks again for your time on discord.

Reply:
Thanks @lordlinus. Rest everything looks good to me

Reply:
> Thanks @ElliotWood
> 
> * Redis is just a cache example and i am planning to have a follow up on how to take those interim messages and return to the user. Autogen output generally take longer and i want to ensure we have a way to display the interim messages if required.
> * In the current use-cases there is no need to have history and hence i did not want to over complicate. I am returning `ChatResult` object and getting `summary` field. You can get `chat_history` and other fields too.. need to change from `res.summary` to `res`.
>   For the next PR i want to include Group chat with Human interaction and Include history and show end to end. Let me know your thoughts
> 
> ![image](https://private-user-images.githubusercontent.com/2416507/323209446-ca9e014a-efc0-43c9-911f-a7f78217d279.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzODAwMTUsIm5iZiI6MTcxMzM3OTcxNSwicGF0aCI6Ii8yNDE2NTA3LzMyMzIwOTQ0Ni1jYTllMDE0YS1lZmMwLTQzYzktOTExZi1hN2Y3ODIxN2QyNzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMTg0ODM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjczOTM3ZjcxYjgxNzVjZmMxNjYxZTY0Y2VjODUxZjU2NjM2MzY0ZjliNmE5MzI4MjIyNjVmMGE4ZGE3M2ZlMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.f0a2WgVP69xHW1kdiORLtu2FKiCDpG8z6zvesub40vg)
> 
> ![image](https://private-user-images.githubusercontent.com/2416507/323209706-1f57ea3d-f562-44ba-92ce-d2927e56f94f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzODAwMTUsIm5iZiI6MTcxMzM3OTcxNSwicGF0aCI6Ii8yNDE2NTA3LzMyMzIwOTcwNi0xZjU3ZWEzZC1mNTYyLTQ0YmEtOTJjZS1kMjkyN2U1NmY5NGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMTg0ODM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDU0MDc2MjE1Y2Y1ZjY4NWYzNmNkZWJjNmQwYTFhNjBlMjhkNTQwZTE4OGI0YmQ2MjU3M2JkYmZhN2E4MzIzNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.6iZBkNGOyXcoU_fklX49BL2nAfL7iYvXWneKLTP9z9s)
> 
> ![image](https://private-user-images.githubusercontent.com/2416507/323209509-02b4a13c-0aae-49fa-8828-d6cbf06ded82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzODAwMTUsIm5iZiI6MTcxMzM3OTcxNSwicGF0aCI6Ii8yNDE2NTA3LzMyMzIwOTUwOS0wMmI0YTEzYy0wYWFlLTQ5ZmEtODgyOC1kNmNiZjA2ZGVkODIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMTg0ODM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjI5YmZmNjE4YzQyZGUyODE2NjI3NTczZjU1Y2RmZTFiNmE3NDcxMmNmZGRmYjEyYzhjMWQ4ZTM4OWU2OTQyZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.GCxsDaKdjA1a5cv-bxYjVlaKm-TExt0E5uDl9JPjtE0)

No problem - you can do something like this.

```
def _chat_history_to_openai(self, chat_history: list) -> list:
        """
            takes the chat history as produced by prompt flow and formats it for use in the simulator
            promptflow uses the following format:
                chat_history = [{"inputs":{"question":"What did OpenAI announce on March 2023?"},
                                 "outputs":{"answer":"On March 2023, OpenAI announced the release of GPT-4"}}]
            simulator uses the following format used by openai chat models:
                [{"role": "human", "content": "What did OpenAI announce on March 2023?"}, 
                 {"role": "agent", "content": "On March 2023, OpenAI announced the release of GPT-4"}]
        """
        formatted_chat_history = []
        for i in chat_history:
            formatted_chat_history.append({"role": "user", "content": i["inputs"][self.question]})
            formatted_chat_history.append({"role": "assistant", "content": i["outputs"][self.answer]})
        return formatted_chat_history
```

```
def _chat_history_to_pf(self, chat_history: list) -> list:
        """
            takes the chat history as produced by openai and converts it to the prompt flow and format
            simulator uses the following format used by openai chat models:
                [{"role": "user", "content": "What did OpenAI announce on March 2023?"}, 
                 {"role": "assistant", "content": "On March 2023, OpenAI announced the release of GPT-4"}]
            promptflow uses the following format:
                chat_history = [{"inputs":{"question":"What did OpenAI announce on March 2023?"},
                                 "outputs":{"answer":"On March 2023, OpenAI announced the release of GPT-4", "context":"some context for the prior answer"}}]
        """
        chat_history = chat_history.copy()
        # filter out any messages that are not from the user or assistant
        chat_history = [message for message in chat_history if message["role"] in ["user", "assistant"]]
        pf_chat_history = []
        for i in range(0, len(chat_history)-1, 2):
            pf_chat_history.append({"inputs":{self.question: chat_history[i]["content"]},
                                    "outputs":{self.answer: chat_history[i+1]["content"]}})
        return pf_chat_history
```